pkg/provider/terraform/instance/apply.go

```diff
@@ -18,30 +23,105 @@ func (p *plugin) terraformApply() error {
 	defer p.applyLock.Unlock()
 
 	if p.applying {
+		select {
+		case p.pollChannel <- true:
+			log.Infoln("Successfully interrupted terraform apply goroutine")
+		default:
+			log.Infoln("Polling channel is full, not interrupting")
+		}
 		return nil
 	}
 
+	p.pollChannel = make(chan bool, 1)
 	go func() {
+		initial := true
 		for {
-			if err := p.lock.TryLock(); err == nil {
-				defer p.lock.Unlock()
-				doTerraformApply(p.Dir)
-			} else {
-				log.Debugln("Can't acquire lock, waiting")
+			if !p.continuePolling() {
+				p.applyLock.Lock()
+				p.applying = false
+				p.applyLock.Unlock()
+				return
+			}
+
+			attempted, err := p.doTerraformApply(initial)
+			initial = false
+			if err != nil {
+				log.Errorf("Executing 'terraform apply' failed: %v", err)
+			}
+			if !attempted {
+				log.Infof("Can't acquire apply lock, waiting %v seconds", p.pollInterval)
+			}
+
+			select {
+			case <-p.pollChannel:
+				// Interrupted, use same initial delay so that more than a single delta
+				// can be processed
+				initial = true
+				break
+			case <-time.After(p.pollInterval):
+				break
 			}
-			time.Sleep(time.Duration(int64(rand.NormFloat64())%1000) * time.Millisecond)
 		}
 	}()
+
 	p.applying = true
 	return nil
 }
 
-func doTerraformApply(dir string) error {
-	log.Infoln(time.Now().Format(time.RFC850) + " Applying plan")
-	command := exec.Command(`terraform apply`).InheritEnvs(true).WithDir(dir)
-	err := command.WithStdout(os.Stdout).WithStderr(os.Stdout).Start()
+// doTerraformApply executes "terraform apply" if it can aquire the lock. Returns
+// true/false if the command was executed and an error
+func (p *plugin) doTerraformApply(initial bool) (bool, error) {
+	if err := p.lock.TryLock(); err == nil {
+		defer p.lock.Unlock()
+		// The trigger for the initial apply is typically from a group commit, sleep
+		// for a few seconds so that multiple .tf.json files have time to be created
+		if initial {
+			time.Sleep(time.Second * 5)
+		}
+		log.Infoln("Applying plan")
+		command := exec.Command(`terraform apply`).InheritEnvs(true).WithDir(p.Dir)
+		err := command.WithStdout(os.Stdout).WithStderr(os.Stdout).Start()
+		if err != nil {
+			return true, err
+		}
+		return true, command.Wait()
+	}
+	return false, nil
+}
+
+// continuePolling returns true if polling should continue; this happens if either
+// the plugin is configured to be standalone or if the associated manager plugin
+// is the current leader
+func (p *plugin) continuePolling() bool {
+	// Always continue if standalone
+	if p.standalone {
+		return true
+	}
+	if err := local.Setup(); err != nil {
```

Comment:
This one too... probably the most important to have this in the `NewTerraformPlugin`.   We should try to error out as early as possible (at creation time of the plugin) when things like creating plugin directories, etc. fail.  This will simplify this code because we don't have to worry about conditions that cannot happen by the time this code executes.